# Unit testing basics
In this chapter, you will get introduced to the pytest package and use it to write simple unit tests. You'll run the tests, interpret the test result reports and fix bugs. Throughout the chapter, we will use examples exclusively from the data preprocessing module of a linear regression project, making sure you learn unit testing in the context of data science.

## Manual testing
The function row_to_list(), which you met in the video lesson, has the following expected return values for the arguments listed below.


| Argument       | Expected return value  | Explanation    |
| :------------- | :----------: | -----------: |
|  "2,081\t314,942\n" | ["2,081", "314,942"]   | Correct row format    |
| "\t293,410\n"  | None | Missing area|
| "1,463238,765\n" | None | Missing tab separator|


row_to_list() has been defined and imported for you. Your job is to test the function manually in the IPython console.

While testing manually, notice how many times you have to repeat the same steps! The point is to experience the inefficiency of manual testing.

In [1]:
def row_to_list(row):
    row = row.rstrip()
    separated_entries = row.split("\t")
    if len(separated_entries) == 2:
        return separated_entries
    return None

## Question
Call row_to_list() in the IPython console on the three arguments listed in the table. Do the actual return values match the expected return values listed in the table?

In [2]:
print(row_to_list("2,081\t314,942\n"))
print(row_to_list("\t293,410\n"))
print(row_to_list("1,463238,765\n"))

['2,081', '314,942']
['', '293,410']
None


No. the function returns ["", "293,410"] for the argument "\t293,410\n" instead of the expected return value None.

In [3]:
def row_to_list_bugfix(row):
    row = row.rstrip()
    separated_entries = row.split("\t")
    if len(separated_entries) == 2 and "" not in separated_entries:
        return separated_entries    
    return None

In the last step, you discovered a bug in our implementation of row_to_list(). Good job!

We have implemented a corresponding bug fix in a new function row_to_list_bugfix(). Call row_to_list_bugfix() in the IPython console on the three arguments listed in the table. Do the actual return values now match the expected return values listed in the table?

In [4]:
print(row_to_list_bugfix("2,081\t314,942\n"))
print(row_to_list_bugfix("\t293,410\n"))
print(row_to_list_bugfix("1,463238,765\n"))

['2,081', '314,942']
None
None


Yes, the implementation returns the expected value in each case.

Well done! Did you notice how manual testing involves repeating the same steps over and over in the IPython console? In this exercise, you just went through a single bug discovery and fixing phase. Just imagine doing this a hundred times over the entire life cycle of row_to_list(), including new feature implementation and refactoring phases! Unit testing can automate these repetitive steps, so that testing becomes easier, and you will learn it in the next lesson ;-)

## Your first unit test using pytest
The data file containing housing area and prices uses commas as thousands separators, e.g. "2,081" or "314,942", as you can see in the IPython Shell.

The convert_to_int() function takes a comma separated integer string as argument, and returns the integer. Therefore, the expected return value of convert_to_int("2,081") is the integer 2081.

This function is defined in the module preprocessing_helpers.py. But it is not known if the function is working properly.

In [5]:
def convert_to_int(number_with_commas):
    return number_with_commas.replace(",", "")

In [6]:
# Import the pytest package
import pytest

# Import the function convert_to_int()
from utils.preprocessing_helpers import convert_to_int

# Complete the unit test name by adding a prefix
def test_on_string_with_one_comma():
  # Complete the assert statement
  assert convert_to_int("2,081") == 2081

You just wrote your first unit test using pytest. Congratulations! A unit test takes 5 minutes to write, but saves you many hours in the future.

## Running unit tests
The tests that you wrote in the previous exercise have been written to a test module test_convert_to_int.py. Try running the tests in the IPython console.

What is the correct IPython console command to run the tests in this test module?`

In [7]:
!pytest utils/test_convert_to_int.py

============================= test session starts ==============================
platform darwin -- Python 3.9.1, pytest-5.2.2, py-1.10.0, pluggy-0.13.1
Matplotlib: 3.1.1
Freetype: 2.10.2
rootdir: /Users/alejandro.robles/PycharmProjects/unit-testing-for-data-science-in-python
plugins: mock-1.11.2, mpl-0.10
collected 1 item                                                               

utils/test_convert_to_int.py F                                           [100%]

=================================== FAILURES ===================================
________________________ test_on_string_with_one_comma _________________________

    def test_on_string_with_one_comma():
>     assert convert_to_int("2,081") == 2081
E     AssertionError: assert '2081' == 2081
E      +  where '2081' = convert_to_int('2,081')

utils/test_convert_to_int.py:10: AssertionError
============================== 1 failed in 0.18s ===============================


Congratulations! You just wrote and ran your first unit test. You probably saw that running the test produced a lot of output. This output contains information about a bug in convert_to_int(). You should be able to read and understand this output in order to bug fix efficiently. This will be covered in the next video lesson. Jump right in!

## What causes a unit test to fail?
In the test result report, the character ., as shown below, stands for a passing test. A passing test is good news as it means that your function works as expected. The character F stands for a failing test. A failing test is bad news as this means that something is broken.

test_row_to_list.py .F.                                                  [100%]
Which of the following describes best why a unit test fails?

### Answer: 
An exception is raised when running the unit test. This could be an AssertionError raised by the assert statement or another exception, e.g. NameError, which is raised before the assert statement can run.

Exactly! If you get an AssertionError, this means the function has a bug and you should fix it. If you get another exception, e.g. NameError, this means that something else is wrong with the unit test code and you should fix it so that the assert statement can actually run.

## Spotting and fixing bugs
To find bugs in functions, you need to follow a four step procedure.

1. Write unit tests.
2. Run them.
3. Read the test result report and spot the bugs.
4. Fix the bugs.

In a previous exercise, you wrote a unit test for the function convert_to_int(), which is supposed to convert a comma separated integer string like "2,081" to the integer 2081. You also ran the unit test and discovered that it is failing.

In this exercise, you will read the test result report from that exercise in detail, and then spot and fix the bug. This would equip you with all basic skills to start using unit tests for your projects.

The convert_to_int() function is defined in the file preprocessing_helpers.py. The unit test is available in the test module test_convert_to_int.py.

## Question
Run the unit test in the test module test_convert_to_int.py in the IPython console. Read the test result report and spot the bug.

Which of the following describes the bug in the function convert_to_int(), if any?

In [8]:
!pytest utils/test_convert_to_int.py

============================= test session starts ==============================
platform darwin -- Python 3.9.1, pytest-5.2.2, py-1.10.0, pluggy-0.13.1
Matplotlib: 3.1.1
Freetype: 2.10.2
rootdir: /Users/alejandro.robles/PycharmProjects/unit-testing-for-data-science-in-python
plugins: mock-1.11.2, mpl-0.10
collected 1 item                                                               

utils/test_convert_to_int.py F                                           [100%]

=================================== FAILURES ===================================
________________________ test_on_string_with_one_comma _________________________

    def test_on_string_with_one_comma():
>     assert convert_to_int("2,081") == 2081
E     AssertionError: assert '2081' == 2081
E      +  where '2081' = convert_to_int('2,081')

utils/test_convert_to_int.py:10: AssertionError
============================== 1 failed in 0.18s ===============================



convert_to_int("2,081") is expected to return the integer 2081, but it is actually returning the string "2081".

Fix the convert_to_int() function so that it returns the integer 2081 instead of the string "2081" for the argument "2,081".

In [9]:
def convert_to_int(string_with_comma):
    # Fix this line so that it returns an int, not a str
    return int(string_with_comma.replace(",", ""))

Good work! Your boss and colleagues are going to really appreciate your new skill of reading test result reports, and then spotting and fixing the bugs, because this would mean fewer bugs in the code base in the long term.

## Unit tests as documentation
Assume that you are a new collaborator of our linear regression project on housing area and prices.

While inspecting the project, you come across a function mystery_function() in the feature module. You want to figure out what this function does. As you know, reading the unit tests might give you the answer quickly!

The unit tests for the function is available in the test module test_mystery_function.py. You can read it, and any other file that you encounter, by using the !cat command in the IPython shell.

Having read the unit tests, can you guess what mystery_function() does?

In [10]:
!cat utils/test_mystery_function.py

import numpy as np
import pytest

from mystery_function import mystery_function

def test_on_clean_data():
    assert np.array_equal(mystery_function("example_clean_data.txt", num_columns=2), np.array([[2081.0, 314942.0], [1059.0, 186606.0]]))

It converts data in a data file into a NumPy array.

You guessed it right and you didn't even take a look at the function definition! This is why - when onboarding new colleagues - it is a good idea to tell them to look at the unit tests if they are not sure about a function's purpose. In Chapter 2, you will see more functions from the feature and models module, and write more advanced unit tests using new pytest features.